core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -242,10 +241,177 @@ public void messageReceived(final ReplicaRequest request, final TransportChannel
         }
     }
 
+    class PrimaryOperationTransportHandler implements TransportRequestHandler<Request> {
+
+        @Override
+        public void messageReceived(Request request, final TransportChannel channel) throws Exception {
+            ActionListener<Response> actionListener = new ActionListener<Response>() {
+                @Override
+                public void onResponse(Response response) {
+                    try {
+                        channel.sendResponse(response);
+                    } catch (Throwable t) {
+                        onFailure(t);
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    try {
+                        channel.sendResponse(e);
+                    } catch (Throwable t) {
+                        logger.warn("failed to send response for get", t);
+                    }
+                }
+            };
+            new AsyncPrimaryAction(request, actionListener).run();
+        }
+    }
+
+    class AsyncPrimaryAction extends AbstractRunnable {
+
+        private final ClusterStateObserver observer;
+        private final ActionListener<Response> listener;
+        private final InternalRequest internalRequest;
+        private final Request request;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
+        private volatile Releasable indexShardReference;
+
+        public AsyncPrimaryAction(Request request, ActionListener<Response> listener) {
+            this.observer = new ClusterStateObserver(clusterService, request.timeout(), logger);
+            this.listener = listener;
+            this.request = request;
+            this.internalRequest = new InternalRequest(request);
+            this.internalRequest.concreteIndex(request.internalShardRouting.getIndex());
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            finishAsFailed(t);
```

Comment:
Just chatted with @brwe and delaying the write request upto the timeout isn't the end of the world in this situation and this situation is rare too. It is much better then the current behaviour. (redirecting the write request endlessly between nodes until the timeout expired)